File: python/src/methods.cpp
#include "bindings.h"

#include <chargefw/methods/method_applicability.h>
#include <chargefw/methods/method_prerequisites.h>
#include <chargefw/methods/method_registry.h>

#include <nanobind/nanobind.h>
#include <nanobind/stl/string.h>

#include <string>

namespace nb = nanobind;

namespace chargefw::python {
namespace {

auto option_value(const methods::MethodOptionValue& value) -> nb::object {
    return std::visit([](const auto& item) { return nb::cast(item); }, value);
}

auto method_descriptors() -> nb::list {
    auto result = nb::list{};
    for (const auto& method : methods::method_registry().methods()) {
        const auto& metadata = method->metadata();
        const auto requirements = method->requirements();
        auto descriptor = nb::dict{};
        descriptor["id"] = std::string{metadata.id};
        descriptor["name"] = std::string{metadata.name};
        descriptor["full_name"] = std::string{metadata.full_name};
        descriptor["publication"] = metadata.publication.has_value()
                                        ? nb::cast(std::string{*metadata.publication})
                                        : nb::none();
        descriptor["priority"] = metadata.priority;
        descriptor["requires_coordinates"] = requirements.coordinates;
        descriptor["supports_cutoff"] = requirements.resources.supports_cutoff;
        descriptor["supports_cover"] = requirements.resources.supports_cover;
        auto options = nb::list{};
        for (const auto& spec : method->option_schema()) {
            auto option = nb::dict{};
            option["id"] = std::string{spec.id};
            option["description"] = std::string{spec.description};
            option["type"] = std::string{methods::to_string(spec.type)};
            option["default"] = option_value(spec.default_value);
            auto choices = nb::list{};
            for (const auto& choice : spec.choices) {
                choices.append(option_value(choice));
            }
            option["choices"] = std::move(choices);
            option["minimum"] = spec.minimum.has_value() ? option_value(*spec.minimum) : nb::none();
            option["minimum_inclusive"] = spec.minimum_inclusive;
            option["maximum"] = spec.maximum.has_value() ? option_value(*spec.maximum) : nb::none();
            option["maximum_inclusive"] = spec.maximum_inclusive;
            options.append(std::move(option));
        }
        descriptor["options"] = std::move(options);
        result.append(std::move(descriptor));
    }
    return result;
}

} // namespace

void bind_methods(nb::module_& module) {
    module.def("_method_descriptors", &method_descriptors);
}

} // namespace chargefw::python

File: python/chargefw/_methods.py
"""Method applicability and execution report values."""

from __future__ import annotations

from dataclasses import dataclass
from typing import cast

from ._catalog import _Catalog
from ._chargefw import methods as _native_methods
from ._parameters import ParameterSetCatalog
from ._payloads import MethodDescriptorPayload
from ._types import (
    ExecutionAvailability,
    ExecutionIssueKind,
    ExecutionMode,
    MethodOptionType,
    PrerequisiteIssueKind,
)


@dataclass(frozen=True, slots=True)
class PrerequisiteIssue:
    kind: PrerequisiteIssueKind
    message: str
    molecule_index: int | None = None
    atom_index: int | None = None
    bond_index: int | None = None
    conformer_index: int | None = None


@dataclass(frozen=True, slots=True)
class ExecutionIssue:
    kind: ExecutionIssueKind
    message: str
    molecule_index: int | None = None


@dataclass(frozen=True, slots=True)
class ExecutionSupport:
    mode: ExecutionMode
    availability: ExecutionAvailability
    issues: tuple[ExecutionIssue, ...] = ()


@dataclass(frozen=True, slots=True, repr=False)
class MethodOption:
    id: str
    description: str
    type: MethodOptionType
    default: bool | int | float | str
    choices: tuple[bool | int | float | str, ...]
    minimum: bool | int | float | str | None
    minimum_inclusive: bool
    maximum: bool | int | float | str | None
    maximum_inclusive: bool

    def __repr__(self) -> str:
        return f"{type(self).__name__}(id={self.id!r}, default={self.default!r})"


class MethodOptionCatalog(_Catalog[MethodOption]):
    """Immutable ordered method options with lookup by ID."""

    __slots__ = ()

    def __init__(self, values: tuple[MethodOption, ...]) -> None:
        super().__init__(values, "method option")


@dataclass(frozen=True, slots=True, repr=False)
class Method:
    id: str
    name: str
    full_name: str
    publication: str | None
    priority: int
    requires_coordinates: bool
    supports_cutoff: bool
    supports_cover: bool
    options: MethodOptionCatalog
    parameter_sets: ParameterSetCatalog

    def __repr__(self) -> str:
        return f"{type(self).__name__}(id={self.id!r}, name={self.name!r})"


class MethodCatalog(_Catalog[Method]):
    """Immutable ordered built-in methods with lookup by ID."""

    __slots__ = ()

    def __init__(self, values: tuple[Method, ...]) -> None:
        super().__init__(values, "method")


def _method_descriptor(
    value: MethodDescriptorPayload, parameter_sets: ParameterSetCatalog
) -> Method:
    return Method(
        id=value["id"],
        name=value["name"],
        full_name=value["full_name"],
        publication=value["publication"],
        priority=value["priority"],
        requires_coordinates=value["requires_coordinates"],
        supports_cutoff=value["supports_cutoff"],
        supports_cover=value["supports_cover"],
        options=MethodOptionCatalog(
            tuple(
                MethodOption(
                    id=option["id"],
                    description=option["description"],
                    type=cast(MethodOptionType, option["type"].name.lower()),
                    default=option["default"],
                    choices=tuple(option["choices"]),
                    minimum=option["minimum"],
                    minimum_inclusive=option["minimum_inclusive"],
                    maximum=option["maximum"],
                    maximum_inclusive=option["maximum_inclusive"],
                )
                for option in value["options"]
            )
        ),
        parameter_sets=parameter_sets.for_method(value["id"]),
    )


def _method_catalog(parameter_sets: ParameterSetCatalog) -> MethodCatalog:
    """Build method descriptors linked to the bundled parameter catalog."""

    return MethodCatalog(
        tuple(
            _method_descriptor(value, parameter_sets)
            for value in _native_methods._method_descriptors()
        )
    )


for _value_type in (
    PrerequisiteIssue,
    ExecutionIssue,
    ExecutionSupport,
    MethodOption,
    MethodOptionCatalog,
    Method,
    MethodCatalog,
):
    _value_type.__module__ = "chargefw"
